src/meta/meta_ckpt.c

```diff
@@ -1270,31 +1416,69 @@ int
 __wt_meta_ckptlist_set(
   WT_SESSION_IMPL *session, WT_DATA_HANDLE *dhandle, WT_CKPT *ckptbase, WT_LSN *ckptlsn)
 {
+    WT_BLOCK_MODS *blk;
     WT_CKPT *ckpt;
     WT_DECL_ITEM(buf);
     WT_DECL_RET;
-    const char *fname;
-    bool has_lsn;
+    WT_ITEM file_blkmods_buffer, checkpoint_blkmods_buffer;
+    u_int i;
+    const char *filename;
+    bool blkmods_are_ok, has_lsn;
 
-    fname = dhandle->name;
-    WT_RET(__wt_scr_alloc(session, 1024, &buf));
+    filename = dhandle->name;
+    blkmods_are_ok = true;
+
+    WT_ERR(__wt_scr_alloc(session, 1024, &buf));
     WT_ERR(__wt_meta_ckptlist_to_meta(session, ckptbase, buf));
     /* Add backup block modifications for any added checkpoint. */
     WT_CKPT_FOREACH (ckptbase, ckpt)
-        if (F_ISSET(ckpt, WT_CKPT_ADD))
+        if (F_ISSET(ckpt, WT_CKPT_ADD)) {
+            WT_CLEAR(file_blkmods_buffer);
+            WT_CLEAR(checkpoint_blkmods_buffer);
+
             WT_ERR(__wt_ckpt_blkmod_to_meta(session, buf, ckpt));
 
+            if (strcmp(filename, WT_HS_URI) == 0 || session->hs_cursor_counter != 0) {
```

Comment:
We need to avoid hitting the following assert in __wt_open_cursor():

Reply:
Are we losing information by skipping files in this situation?  That is, `_wt_meta_ckptlist_set` on file X can apparently be called while a history store cursor is open.  If we skip either the collection of bits or checking of bits when that happens, is that a bug or just unfortunate?

Since we are opening a metadata cursor (readonly presumably) that should be safe, as the metadata doesn't participate in history store.  So I think we could change `__wt_open_cursor` as:
```
     * There are three exceptions to this rule:
     *  - Verifying the metadata through an internal session.
     *  - The btree is being verified.
     *  - Opening the metadata file itself.       /* ADDED */
     */
    WT_ASSERT(session,
      strcmp(uri, WT_HS_URI) == 0 || 
      strcmp(uri, WT_METADATA_URI) == 0 ||    /* ADDED */
      session->hs_cursor_counter == 0 ||
        F_ISSET(session, WT_SESSION_INTERNAL) ||
        (S2BT_SAFE(session) != NULL && F_ISSET(S2BT(session), WT_BTREE_VERIFY)));
```
@etienneptl added this assert in WT-6682 about 8 months ago.  Etienne, would this be a safe change?

Reply:
Yes, @ddanderson, I believe this is safe - however, can we ensure the metadata cursor is read-only?

Reply:
@jeremythorp, assuming there's a way to fix the assert check in `__wt_open_cursor` so it wouldn't fire, is the code improved by removing `session->hs_cursor_counter != 0` check?  Do we check more cases than we would have?

If so, then we could pursue fixing the assert.   @etienneptl, there could be a way to ensure this is a readonly cursor if that is important for correctness. I.e. (would something bad happen if it were not a readonly metadata cursor).  It's a little awkward to check for readonly-ness because at the point of the WT_ASSERT in `__wt_open_cursor`, there are ways to do it, but add a little complexity.  Let's call this "complex assert".

So I think there are three ways out 

1.  keep the code in the PR (maybe it doesn't check some cases we should??), leave cursor_open alone
2. tighten up the ASSERT in meta_ckpt and loosen the ASSERT in cursor open as suggested in code in my comment (maybe it doesn't protect against writeable metadata cursor opens, is that important??)
3. tighten up the ASSERT in meta_ckpt, and use the "complex assert" (I'll describe in another comment).

Reply:
The "complex assert" would move first move the assert to the end of `__wt_open_cursor`.  So there would be an `err:` label and all the returns (including return(0)) would jump to it.
```
err:
#ifdef WT_DIAGNOSTIC
     if (ret == 0)
         readonly = WT_CURSOR_IS_READONLY(*cursorp);
     else
#define WT_CURSOR_IS_READONLY(cursor)        ((cursor)->insert == __wt_cursor_notsup)

Reply:
Hi @ddanderson , I don't like the idea of the added complexity either, so I'd prefer an alternative. 
@etienneptl , do you have any thoughts on this? Thanks, Jeremt

Reply:
@ddanderson, you said:

> That is, _wt_meta_ckptlist_set on file X can apparently be called while a history store cursor is open

Do we know why we have a HS cursor open in this code flow now? I am not understanding this bit.

Reply:
@etienneptl, 
> Do we know why we have a HS cursor open in this code flow now? I am not understanding this bit.

I'd guess that this is a checkpoint thread, and perhaps is in the middle of reconciliation, which will need to update the history store.

Reply:
The questions I have now:

- If we go with @jeremythorp's suggestion, we skip the HS table given the comment line 1482. Does it mean we skip that table while we should process it?
- If we update the other assertion, does it means we process the HS and that's what we want?

Reply:
After re-reading the code, here is my understanding:

- If we skip the HS, we are not validating the blocks associated with it which imo is wrong. If they are invalid and don't realise it, if the backup is used, it will fail at restoring the right version of the HS. I am against skipping the HS file.
- The new code from the PR does the following as part of checkpoint (below is a stack trace from a failure):

```
__txn_checkpoint
__checkpoint_apply_to_dhandles
__checkpoint_tree_helper
__checkpoint_tree
__wt_meta_ckptlist_set ---> Updated function where we have an issue with an assertion down the road
__check_backup_blocks
__get_blkmods
__wt_metadata_cursor_open
__wt_open_cursor
```

The checkpoint does open a hs cursor before getting to `__wt_meta_ckptlist_set`. We can see that at the end, we open a metadata cursor file. As pointed out by @ddanderson [here](https://github.com/wiredtiger/wiredtiger/pull/9437/files#r1410814813), the assertion should be updated so it allows that code flow when we are opening a new cursor on the metadata file despite having a HS cursor already. I wonder if we should allow this only when a checkpoint is running?

Reply:
Hi @ddanderson , yes this is the main issue holding up the PR, so that's why I'm keen to resolve it. @etienneptl and and I are investigating what it would take to make the code work if we remove the 'skip history store' step from the checks - we'll keep you posted.

Reply:
Hi @etienneptl and @ddanderson , I've run a series of patch builds with the History Store check removed. I saw a number of failures in the assert in session_api.c. The details are in this [JIRA comment](https://jira.mongodb.org/browse/WT-11178?focusedCommentId=6180603&focusedId=6180603&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6180603).

Reply:
@etienneptl , thanks very much for your help. The next step is to adjust the ASSERT to be ok if we have both a metadata cursor AND a checkpoint is running. I'll make the change tomorrow, test it, and let you know what happens.

Reply:
@jeremythorp and I had a discussion and we agreed on updating the assertion with the following:
```
diff --git a/src/session/session_api.c b/src/session/session_api.c
index 246c7e879..d4c4f08e0 100644
--- a/src/session/session_api.c
+++ b/src/session/session_api.c
@@ -693,10 +693,13 @@ __wt_open_cursor(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner, co
      * There are two exceptions to this rule:
      *  - Verifying the metadata through an internal session.
      *  - The btree is being verified.
+     *  - Opening the metadata file itself while performing a checkpoint.
      */
     WT_ASSERT(session,
-      strcmp(uri, WT_HS_URI) == 0 || session->hs_cursor_counter == 0 ||
-        F_ISSET(session, WT_SESSION_INTERNAL) ||
+      strcmp(uri, WT_HS_URI) == 0 ||
+        (strcmp(uri, WT_METADATA_URI) == 0 &&
+          __wt_atomic_loadvbool(&txn_global->checkpoint_running)) ||
+        session->hs_cursor_counter == 0 || F_ISSET(session, WT_SESSION_INTERNAL) ||
         (S2BT_SAFE(session) != NULL && F_ISSET(S2BT(session), WT_BTREE_VERIFY)));
 
     /* We do not cache any subordinate tables/files cursors. */
```

Reply:
Sounds good.

Reply:
Looks like the check is wrong, it should be `WT_METAFILE_URI` instead.

Reply:
Now fixed!